hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java

```diff
@@ -1687,6 +1731,58 @@ public MerkleDiskFs diskFs() {
 		return state.diskFs();
 	}
 
+	/**
+	 * return the directory to which record stream files should be write
+	 *
+	 * @return the direct file folder for writing record stream files
+	 */
+	public String getRecordStreamDirectory() {
+		if (recordStreamDir == null) {
+			final String nodeAccountString = EntityIdUtils.asLiteralString(nodeAccount());
+			String parentDir = properties().getStringProperty("hedera.recordStream.logDir");
+			if (!parentDir.endsWith(File.separator)) {
+				parentDir += File.separator;
+			}
+			recordStreamDir = parentDir + "record" + nodeAccountString;
+		}
+		return recordStreamDir;
+	}
+
+	/**
+	 * update the runningHash instance saved in runningHashLeaf
+	 *
+	 * @param runningHash
+	 * 		new runningHash instance
+	 */
+	public void updateRecordRunningHash(final RunningHash runningHash) {
+		state.runningHashLeaf().setRunningHash(runningHash);
+	}
+
+	/**
+	 * set recordsInitialHash, which will be set to RecordStreamManager as initialHash.
+	 * recordsInitialHash is read at restart, either from the state's runningHashLeaf,
+	 * or from the last old .rcd_sig file in migration.
+	 * When recordsInitialHash is read, the RecordStreamManager might not be initialized yet,
+	 * because RecordStreamManager can only be initialized after system files are loaded so that enableRecordStream
+	 * setting is read.
+	 * Thus we save the initialHash in the context, and use it when initializing RecordStreamManager
+	 *
+	 * @param recordsInitialHash
+	 * 		initial running Hash of records
+	 */
+	public void setRecordsInitialHash(final Hash recordsInitialHash) {
+		if (recordsInitialHash != null) {
```

Comment:
If there are two hash values here, how to decide which one to keep?

Reply:
yes, agreed. Removed this null check. Thanks @ljianghedera